[WebXR Hit Test] imported/w3c/web-platform-tests/webxr/hit-test/ar_hittest_subscription_refSpaces.https.html fails

fujii opened the pull request:
#### f8e8db847038a190ef76814e16ba9c57ace77e24
<pre>
[WebXR Hit Test] imported/w3c/web-platform-tests/webxr/hit-test/ar_hittest_subscription_refSpaces.https.html fails
<a href="https://bugs.webkit.org/show_bug.cgi?id=303516">https://bugs.webkit.org/show_bug.cgi?id=303516</a>

Reviewed by Sergio Villar Senin.

Take the space of XRHitTestOptionsInit into account. Hit test should be
performed in the space. Added a new type PlatformXR::NativeOriginInformation
that supports reference spaces and input source spaces at the moment.

Added a new WebXRSpace subclass WebXRHitTestResultSpace in
WebXRHitTestResult.cpp to populate the pose of hit test result.

Tests: imported/w3c/web-platform-tests/webxr/hit-test/ar_hittest_subscription_refSpaces.https.html
       imported/w3c/web-platform-tests/webxr/hit-test/ar_hittest_subscription_inputSources.https.html

* LayoutTests/imported/w3c/web-platform-tests/webxr/hit-test/ar_hittest_subscription_inputSources.https-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/webxr/hit-test/ar_hittest_subscription_refSpaces.https-expected.txt:
* Source/WebCore/Modules/webxr/WebXRHitTestResult.cpp:
(WebCore::WebXRHitTestResultSpace::WebXRHitTestResultSpace):
(WebCore::WebXRHitTestResult::getPose):
* Source/WebCore/Modules/webxr/WebXRInputSource.cpp:
(WebCore::WebXRInputSource::WebXRInputSource):
(WebCore::WebXRInputSource::update):
* Source/WebCore/Modules/webxr/WebXRInputSpace.cpp:
(WebCore::WebXRInputSpace::create):
(WebCore::WebXRInputSpace::WebXRInputSpace):
(WebCore::WebXRInputSpace::nativeOriginInformation const):
* Source/WebCore/Modules/webxr/WebXRInputSpace.h:
* Source/WebCore/Modules/webxr/WebXRReferenceSpace.h:
* Source/WebCore/Modules/webxr/WebXRSession.cpp:
(WebCore::WebXRSession::requestHitTestSource):
* Source/WebCore/Modules/webxr/WebXRSpace.h:
(WebCore::WebXRSpace::nativeOriginInformation const):
* Source/WebCore/platform/xr/PlatformXR.h:
* Source/WebCore/testing/WebFakeXRDevice.cpp:
(WebCore::SimulatedXRDevice::frameTimerFired):
* Source/WebKit/Shared/XR/PlatformXR.serialization.in:

Canonical link: <a href="https://commits.webkit.org/304264@main">https://commits.webkit.org/304264@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/cd5ccd77cea683008185ca78ea96afe0585cc67a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/135054 "Passed style check") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/159/builds/7456 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/46306 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/142561 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/86880 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/136923 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/8084 "Built successfully") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/156/builds/7305 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/103199 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/86880 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/138000 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/162/builds/5738 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/121045 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/84052 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/154/builds/5554 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [⏳ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/macOS-Tahoe-Debug-API-Tests-EWS "Waiting to run tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/3157 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/114763 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/39204 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/145259 "Built successfully") | 
| | [  ~~🛠 vision~~](https://ews-build.webkit.org/#/builders/153/builds/7136 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/136/builds/39779 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/111572 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/160/builds/7185 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/156/builds/7305 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/111935 "Passed tests") | 
| | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-26-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/117331 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/61076 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/20834 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/7186 "Built successfully") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/35496 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/157/builds/6954 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/70757 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/163/builds/7185 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/152/builds/7060 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

svillar commented on Source/WebKit/Shared/XR/PlatformXR.serialization.in:
```diff
@@ -22,6 +22,15 @@
+    LocalFloor,
+    BoundedFloor,
+    Unbounded,
+    Webgpu
```
Webgpu as reference space type?

fujii replied:
This definition in *.serialization.in has to match with the original enum definition.
I think we can remove ReferenceSpaceType::Webgpu. I will do it later.

fujii replied:
Fixed by https://commits.webkit.org/304189@main

svillar commented on Source/WebCore/Modules/webxr/WebXRHitTestResult.cpp:
```diff
@@ -55,7 +55,7 @@ WebXRHitTestResult::~WebXRHitTestResult() = default;
 {
     PlatformXR::FrameData::InputSourcePose pose { m_result.pose };
-    Ref space = WebXRInputSpace::create(document, m_frame->session(), pose);
+    Ref space = WebXRInputSpace::create(document, m_frame->session(), pose, 0);
```
why always 0 ?

How can that handle be another value?

fujii replied:
I think I should add a new XRSpace subclass for this. I will do it as a follow-up.

svillar replied:
Then I prefer not to add this magic value here. Just initialize it whenever required. You could add the parameter in the follow up. Also please add a FIXME mentioning the new subclass to be added.

fujii replied:
Added a new sublass WebXRHitTestResultSpace. Fixed.

svillar commented on Source/WebCore/Modules/webxr/WebXRReferenceSpace.cpp:
```diff
@@ -98,6 +98,13 @@ std::optional<TransformationMatrix> WebXRReferenceSpace::nativeOrigin() const
+    return { m_type };
+}
+#endif
+
```
Let's add this to the header, it's a simple getter

svillar commented on Source/WebCore/Modules/webxr/WebXRSpace.cpp:
```diff
@@ -72,6 +72,13 @@ ScriptExecutionContext* WebXRSpace::scriptExecutionContext() const
+{
+    return std::nullopt;
+}
+#endif
```
Ditto

svillar reviewed (approved):
r+

I just have 2 minor nits

BTW please also mention the new subclass in the commit message

fujii committed:
[WebXR Hit Test] imported/w3c/web-platform-tests/webxr/hit-test/ar_hittest_subscription_refSpaces.https.html fails
https://bugs.webkit.org/show_bug.cgi?id=303516

Reviewed by Sergio Villar Senin.

Take the space of XRHitTestOptionsInit into account. Hit test should be
performed in the space. Added a new type PlatformXR::NativeOriginInformation
that supports reference spaces and input source spaces at the moment.

Added a new WebXRSpace subclass WebXRHitTestResultSpace in
WebXRHitTestResult.cpp to populate the pose of hit test result.

Tests: imported/w3c/web-platform-tests/webxr/hit-test/ar_hittest_subscription_refSpaces.https.html
       imported/w3c/web-platform-tests/webxr/hit-test/ar_hittest_subscription_inputSources.https.html

* LayoutTests/imported/w3c/web-platform-tests/webxr/hit-test/ar_hittest_subscription_inputSources.https-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/webxr/hit-test/ar_hittest_subscription_refSpaces.https-expected.txt:
* Source/WebCore/Modules/webxr/WebXRHitTestResult.cpp:
(WebCore::WebXRHitTestResultSpace::WebXRHitTestResultSpace):
(WebCore::WebXRHitTestResult::getPose):
* Source/WebCore/Modules/webxr/WebXRInputSource.cpp:
(WebCore::WebXRInputSource::WebXRInputSource):
(WebCore::WebXRInputSource::update):
* Source/WebCore/Modules/webxr/WebXRInputSpace.cpp:
(WebCore::WebXRInputSpace::create):
(WebCore::WebXRInputSpace::WebXRInputSpace):
(WebCore::WebXRInputSpace::nativeOriginInformation const):
* Source/WebCore/Modules/webxr/WebXRInputSpace.h:
* Source/WebCore/Modules/webxr/WebXRReferenceSpace.h:
* Source/WebCore/Modules/webxr/WebXRSession.cpp:
(WebCore::WebXRSession::requestHitTestSource):
* Source/WebCore/Modules/webxr/WebXRSpace.h:
(WebCore::WebXRSpace::nativeOriginInformation const):
* Source/WebCore/platform/xr/PlatformXR.h:
* Source/WebCore/testing/WebFakeXRDevice.cpp:
(WebCore::SimulatedXRDevice::frameTimerFired):
* Source/WebKit/Shared/XR/PlatformXR.serialization.in:

Canonical link: https://commits.webkit.org/304264@main
